docs/architecture/adr-044-lite-client-with-weak-subjectivity.md

```diff
@@ -121,6 +124,176 @@ network usage.
 Check out the formal specification
 [here](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md).
 
+### Implementation
+
+There are two primary modes of usage right now:
+
+1) Trusted RPC proxy (wrapping an RPC client + verification)
+2) Part of the IBC light client (only verification bit, no RPC) [spec](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics)
+
+First, we'll need something, which will provide us secure headers & validator sets.
+
+```go
+type Provider interface {
+	// 0 - latest
+	SignedHeader(height int64) (SignedHeader, error)
+
+  // 0 - latest
+  Validators(height int64) (ValidatorSet, error)
+}
+```
+
+In case of the proxy it will be a `http` provider (wrapping RPC client). For
+IBC, it will be a `ibc` provider, receiving information from IBC transactions.
```

Comment:
I'm not sure there will be something in the blockchain app that is actually going to be calling these functions though? It seems like that's the relayers job, and then the relayer submits all this stuff and once it's recieved by the app it just needs to call verify with whatever it received. As far as I know the app isn't going to be asking anything to "please send me signed header for block 5", it will just try to verify whatever it receives.

So that can/should probably be better clarified.

Reply:
> As far as I know the app isn't going to be asking anything to "please send me signed header for block 5", it will just try to verify whatever it receives.

What about bisection in IBC?

> So that can/should probably be better clarified.

Agree

Reply:
This is very interesting as it seems that part of bisection logic could (should) be implemented by the relayer.